osemgrep: Port autofix application for `fix`

nmote opened the pull request:
This isn't an exact port of the relevant code in `autofix.py`. I already had some autofix application code in `Autofix.ml` for testing purposes. In #6391 I refactored it to make it usable in this context.

Several differences:
- Instead of tracking details about applied fixes so that we can correctly apply multiple fixes to the same file, we just apply them from bottom to top so that they can't interfere with each other.
- We discard overlapping fixes and apply the ones that we can apply without conflicts. I think the Python CLI just applies them anyway, possibly incorrectly. However, the Python CLI also dedups matches that are the same except for metavariable bindings, which eliminates a major source of overlapping fixes. This logic is not ported yet.
- Here, we apply fixes using the byte offset into the file, instead of reading the file, splitting into lines, and then updating using the line and column like the CLI does. This is a lot simpler.
- When the `--dryrun` flag is passed, the Python CLI reports a list of fixed lines in its JSON output. This hasn't been ported yet, and in fact using the byte offset (see above) will make this a little bit more complicated.

Test plan:

Manual testing with the example given in #4964:

`osemgrep scan -c autofix.yaml autofix.py --autofix`

Note that this is an example where there are multiple matches reported for the same range, due to different possible metavariable bindings. For overlapping ranges, osemgrep applies the first one in the list that semgrep-core returns, just like the Python CLI. So, in this example it applies the same fix as the CLI.

Also, ran the e2e tests:

`PYTEST_USE_OSEMGREP=true pipenv run python -m pytest tests/e2e`

Before: 77 passed

After: 83 passed

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

ievans commented:
I wonder if we can just skip the --dryrun flag in the port and maybe add it back later

nmote commented:
I already ported `--dryrun` in this PR, and it was easy enough to do for the part of it that I ported. It's the `fixed_lines` in particular that is going to be a little annoying. It probably won't be too bad, though. I just mentioned it because I chose to put it off for this specific PR. I don't know what it's used for (if anything).

aryx commented:
Woohoo! Agree with the approach. This is often what we did with martin; Looking at the original code, taking the good bits but not being afraid to follow a different approach if it makes things significantly simpler.

aryx commented on semgrep-core/src/fixing/Autofix.ml:
```diff
@@ -94,6 +94,25 @@ let apply_edits_to_text text edits =
 
+let partition_edits_by_file edits =
+  let edits_by_file = Hashtbl.create 8 in
+  List.iter
```
probably can use one of the Common.group_by function to do this.

nmote replied:
Looks like `Common.group_by` reverses the order of the list elements. The order is important here. If I reverse it after calling `group_by`, then I'm relying on `group_by` to always reverse the order of the list elements. If someone changes it later, this will break.

```
let group_by f xs =
  (* use Hashtbl.find_all property *)
  let h = Hashtbl.create 101 in

  (* could use Set *)
  let hkeys = Hashtbl.create 101 in

  xs |> List.iter (fun x ->
    let k = f x in
    Hashtbl.replace hkeys k true;
    Hashtbl.add h k x
  );
  Hashtbl.fold (fun k _ acc -> (k, Hashtbl.find_all h k)::acc) hkeys []
;;

let l = ["1"; "2"; "3"];;

let g = group_by (fun _ -> 5) l;;

(* output: 3, 2, 1 *)
print_endline (g |> List.hd |> snd |> String.concat ", ");;
```

aryx commented on semgrep-core/src/fixing/Autofix.mli:
```diff
@@ -16,6 +16,22 @@ val render_fix :
   string option
 
+type textedit = {
+  path : string;
```
Common.filename.

aryx commented on semgrep-core/src/osemgrep/cli_scan/Output.ml:
```diff
@@ -21,6 +35,25 @@ let output_result (conf : Scan_CLI.conf) (res : Core_runner.result) : unit =
+   let modified_files, _failed_fixes = apply_fixes conf cli_output in
+   if not conf.dryrun then
+     if modified_files <> [] then
+       pr2
```
You should now be able to use the Logs library. See my recent PR regarding logging.

nmote replied:
I tried to faithfully reproduce the Python behavior here. It prints on stderr with no additional text other than what's seen here. I just tried out the Logs library and got:

```
osemgrep: [WARNING] successfully modified 1 file.
```

I had to use `Logs.warn` because otherwise it looks like this wouldn't get printed by default. This output seems weird and it's inconsistent with the existing output.

I guess I'm not sure how faithfully we are trying to reproduce the current behavior. Does it matter whether this is slightly different? If I were writing this from scratch I would probably use `Logs.info` for this, so it appears only when `--verbose` or higher is passed.

Anyway, I'm going to merge this without using Logs here but I'll be happy to make a followup PR depending on your answer here.

aryx reviewed (approved):
Fantastic!

nmote committed:
osemgrep: Port autofix application for `fix`

This isn't an exact port of the relevant code in `autofix.py`. I already
had some autofix application code in `Autofix.ml` for testing purposes.
In #6391 I refactored it to make it usable in this context.

Several differences:
- Instead of tracking details about applied fixes so that we can
  correctly apply multiple fixes to the same file, we just apply them
  from bottom to top so that they can't interfere with each other.
- We discard overlapping fixes and apply the ones that we can apply
  without conflicts. I think the Python CLI just applies them anyway,
  possibly incorrectly. However, the Python CLI also dedups matches that
  are the same except for metavariable bindings, which eliminates a
  major source of overlapping fixes. This logic is not ported yet.
- Here, we apply fixes using the byte offset into the file, instead of
  reading the file, splitting into lines, and then updating using the
  line and column like the CLI does. This is a lot simpler.
- When the `--dryrun` flag is passed, the Python CLI reports a list of
  fixed lines in its JSON output. This hasn't been ported yet, and in
  fact using the byte offset (see above) will make this a little bit
  more complicated.

Test plan:

Manual testing with the example given in #4964:

`osemgrep scan -c autofix.yaml autofix.py --autofix`

Note that this is an example where there are multiple matches reported
for the same range, due to different possible metavariable bindings. For
overlapping ranges, osemgrep applies the first one in the list that
semgrep-core returns, just like the Python CLI. So, in this example it
applies the same fix as the CLI.

Also, ran the e2e tests:

`PYTEST_USE_OSEMGREP=true pipenv run python -m pytest tests/e2e`

Before: 77 passed

After: 83 passed

nmote commented:
CI failure is unrelated
